src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -319,20 +319,30 @@ public static AsyncTaskMethodBuilder<TResult> Create()
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.stateMachine);
             }
 
-            // Run the MoveNext method within a copy-on-write ExecutionContext scope.
-            // This allows us to undo any ExecutionContext changes made in MoveNext,
+            Thread currentThread = Thread.CurrentThread;
+            // Capture references to Thread Contexts
+            ref ExecutionContext current = ref currentThread.ExecutionContext;
+            ref SynchronizationContext currentSyncCtx = ref currentThread.SynchronizationContext;
+
+            // Store current ExecutionContext and SynchronizationContext as "previous"
+            // This allows us to restore them and undo any Context changes made in stateMachine.MoveNext
             // so that they won't "leak" out of the first await.
+            ExecutionContext previous = current;
+            SynchronizationContext previousSyncCtx = currentSyncCtx;
 
-            Thread currentThread = Thread.CurrentThread;
-            ExecutionContextSwitcher ecs = default(ExecutionContextSwitcher);
-            try
+            // Async state machines are required not to throw, so no need for try/finally here.
```

Comment:
corert specifies this [`corert/.../Start<TStateMachine>(ref TStateMachine stateMachine)`](https://github.com/dotnet/corert/blob/843f6ab3477eeb36f221d502e5ceaa780ac6c503/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs#L727-L732); valid approach?

Reply:
Would resolve https://github.com/dotnet/coreclr/issues/11156

Reply:
As described in #11156, this is debatable.  Let's keep PRs focused on one thing rather than trying to cram such changes together.

Reply:
Removed